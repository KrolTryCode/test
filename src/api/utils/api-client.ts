import { AxiosError } from 'axios';
import i18n from 'i18next';
import queryString from 'query-string';

import { ApiConfig, Api } from '~/api/utils/api-requests';
import { authPath, loginPath } from '~/utils/configuration/routes-paths';
import {
  projectLocalStorageService,
  projectSessionStorageService,
} from '~/utils/localstorage/project-storage/project-storage-instance';
import { ProjectStorageKey } from '~/utils/localstorage/project-storage/project-storage.types';
import { showErrorMessage } from '~/utils/show-error-message';

export interface ErrorResponse {
  code: number;
  message: string;
}

const getStorageService = () => {
  const rememberMe = projectLocalStorageService.get(ProjectStorageKey.RememberMe);
  return rememberMe ? projectLocalStorageService : projectSessionStorageService;
};

const BASE_API_URL = '/api';

const API_CONFIG: ApiConfig = {
  baseURL: BASE_API_URL,
};

const API_CONFIG_SECURE: ApiConfig = {
  baseURL: BASE_API_URL,
  securityWorker: async () => {
    const token = getStorageService().get(ProjectStorageKey.AccessToken);

    if (token) {
      return { headers: { Authorization: `Bearer ${token ?? ''}` } };
    }
  },
  paramsSerializer: params => {
    const keys = Object.keys(params);

    let res: {
      [key: string]: unknown;
    } = {};

    for (const key of keys) {
      if (typeof params[key] === 'object' && !Array.isArray(params[key])) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        res = { ...res, ...params[key] };
      } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        res[key] = params[key];
      }
    }

    return queryString.stringify(res, { arrayFormat: 'none' });
  },
};

export const ApiClient = new Api(API_CONFIG);

export const ApiClientSecured = new Api(API_CONFIG_SECURE);

ApiClientSecured.instance.interceptors.request.use(request => {
  request.headers.set('Accept-Language', i18n.language);
  return request;
});

ApiClient.instance.interceptors.request.use(request => {
  request.headers.set('Accept-Language', i18n.language);
  return request;
});

// Можно задать кастомные обработчики для каждой ошибки
ApiClientSecured.instance.interceptors.response.use(
  response => response,
  async (error: AxiosError) => {
    if (error.response?.status === 502) {
      showErrorMessage(error, 'ERROR.TECHNICAL_ERROR');
    }
    if (error.response?.status === 401) {
      if (error.config?.url === '/api/v1/jwt/refresh') {
        window.location = `/${authPath}/${loginPath}` as unknown as Location;
      } else {
        if (window.location.pathname.startsWith(`/${authPath}`)) {
          return Promise.reject(error);
        }
        const storage = getStorageService();
        storage.remove(ProjectStorageKey.AccessToken);
        const refreshToken = storage.get(ProjectStorageKey.RefreshToken);
        const originalRequest = error.config;
        let token = '';

        await ApiClientSecured.tokenV1Controller
          .regenerateAccessToken({ headers: { Authorization: `Bearer ${refreshToken ?? ''}` } })
          .then(res => {
            token = res.token ?? '';
            storage.set(ProjectStorageKey.AccessToken, token);
          });
        return ApiClientSecured.instance.request({
          ...originalRequest,
          headers: { ...originalRequest?.headers, Authorization: `Bearer ${token}` },
        });
      }
    }
    return Promise.reject(error);
  },
);
