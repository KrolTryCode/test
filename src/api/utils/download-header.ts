import { RequestParams } from './api-requests';

export const getDownloadHeaderCSV = (): RequestParams => ({
  headers: {
    accept: 'text/csv',
  },
});
