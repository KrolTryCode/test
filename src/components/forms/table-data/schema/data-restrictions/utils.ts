import {
  ColumnValue,
  Const,
  FloatConst,
  IntConst,
  StringConst,
  StringLengthValue,
} from '~/api/utils/api-requests';

export const isColumnValueType = (value: any) => {
  return '@type' in value && value['@type'] === 'ColumnValue';
};

export const getRightOrLeftValue = (
  currentObject:
    | string
    | object
    | ColumnValue
    | Const
    | StringLengthValue
    | FloatConst
    | IntConst
    | StringConst,
) => {
  let value: object | string;
  if (typeof currentObject === 'object' && 'value' in currentObject) {
    value = currentObject.value;
  } else {
    value = currentObject;
  }
  return value;
};
