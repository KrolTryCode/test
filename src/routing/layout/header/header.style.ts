import { styled } from '@mui/material';
import { Link } from '@tanstack/react-router';

export const StyledHeader = styled('header')(
  ({ theme }) =>
    `
  z-index: 2;
  display: flex;
  flex: 0 0 auto;
  flex-direction: row;
  height: 66px;
  padding: 0 0 0 ${theme.spacing(1)};
  background: ${theme.palette.primary.main};
  box-shadow:
    0 1px 3px rgba(0, 0, 0, 0.12),
    0 1px 2px rgba(0, 0, 0, 0.24);

  img.logo {
    width: 50px;
    height: 50px;
  }
`,
);

export const QuestionMarkWrapper = styled('b')(
  ({ theme }) => `
  display: block;
  width: 20px;
  margin: 8px;
  font-size: 14px;
  color: ${theme.palette.primary.main};
  text-align: center;
  background: white;
  border-radius: 50%;
`,
);

export const StyledHeaderHelp = styled(Link)`
  padding: 8px;
  margin-top: -3px;
  text-decoration: none;
`;
