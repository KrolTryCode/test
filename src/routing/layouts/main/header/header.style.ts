import { styled } from '@mui/material';
import { Link } from 'react-router-dom';

export const StyledHeader = styled('header')(
  ({ theme }) =>
    `
  display: flex;
  flex: 0 0 auto;
  flex-direction: row;
  padding: 0 0 0 ${theme.spacing(1)};
  background: ${theme.palette.primary.main};
  box-shadow:
    0 1px 3px rgba(0, 0, 0, 0.12),
    0 1px 2px rgba(0, 0, 0, 0.24);

  img {
    width: 50px;
    height: 50px;
  }
`,
);

export const QuestionMarkWrapper = styled('b')(
  ({ theme }) => `
  display: block;
  width: 20px;
  height: 20px;
  font-size: 14px;
  color: ${theme.palette.primary.main};
  text-align: center;
  background: white;
  border-radius: 50%;
`,
);

export const StyledHeaderHelp = styled(Link)`
  margin: auto 20px;
  text-decoration: none;
`;
