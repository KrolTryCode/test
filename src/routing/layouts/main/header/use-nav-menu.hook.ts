import { useMemo } from 'react';

import { useGetExternalLinksQuery } from '~/api/queries/settings/get-external-links.query';
import { useUserStore } from '~/app/user/user.store';
import { routes } from '~/routing/routes/routes';
import { RouteDescription } from '~/routing/routes.types';

export const useNavMenu = () => {
  const permissions = useUserStore(store => store.data?.permissions ?? []);

  const { data: externalLinks } = useGetExternalLinksQuery({ select: data => data.links });

  const topMenuRoutes: RouteDescription[] = useMemo(
    () =>
      routes
        .find(({ key }) => key === 'main')
        ?.children?.filter(
          ({ menuDisplay, accessBy }) =>
            !!menuDisplay && (accessBy ? accessBy.some(role => permissions.includes(role)) : true),
        ) ?? [],
    [permissions],
  );

  return { topMenuRoutes, externalLinks };
};
