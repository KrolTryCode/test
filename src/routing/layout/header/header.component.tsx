import { Notifications } from '@mui/icons-material';
import { Badge, Box, useMediaQuery, useTheme } from '@mui/material';
import { useQuery } from '@tanstack/react-query';
import { Link } from '@tanstack/react-router';
import { FC, useMemo } from 'react';
import { useTranslation } from 'react-i18next';

import { getNotificationsListOptions } from '~/api/queries/notifications/get-notifications-list.query';
import { selectUnreadNotificationsCount } from '~/api/selectors/select-notifications';
import { logo } from '~/utils/configuration/design/logo';
import { useAppDesignConfig } from '~/utils/configuration/design/use-app-design-config.hook';
import { createObjectURLFromFile } from '~/utils/files/create-object-url-from-file';
import { useServerPagingParams } from '~/utils/hooks';

import { QuestionMarkWrapper, StyledHeader, StyledHeaderHelp } from './header.style';
import { NavMenuDesktop } from './nav-menu-desktop/nav-menu-desktop.component';
import { NavMenuMobile } from './nav-menu-mobile/nav-menu-mobile.component';
import { UserMenu } from './user-menu/user-menu.component';

const Header: FC = () => {
  const { t } = useTranslation();
  const { appMainLogo } = useAppDesignConfig();
  const { data: unreadCount } = useQuery(
    getNotificationsListOptions(useServerPagingParams(), {
      select: selectUnreadNotificationsCount,
    }),
  );
  const imgSrc = useMemo(() => createObjectURLFromFile(appMainLogo), [appMainLogo]);
  const theme = useTheme();
  const isDesktop = useMediaQuery(theme.breakpoints.up('sm'));

  const DEFAULT_APP_MAIN_LOGO = logo['logoSmall'];

  return (
    <StyledHeader>
      <Box hidden={!isDesktop} marginBlock={'auto'} marginRight={2}>
        <Box
          component={'img'}
          display={'block'}
          src={imgSrc ?? DEFAULT_APP_MAIN_LOGO}
          className={'logo'}
          alt={t('PROJECT_NAME').toString()}
        />
      </Box>

      {isDesktop ? <NavMenuDesktop /> : <NavMenuMobile />}

      <Box
        display={'flex'}
        flexDirection={'row'}
        marginLeft={'auto'}
        alignItems={'center'}
        paddingRight={1}
      >
        <StyledHeaderHelp to={'/Help.pdf'} target={'_blank'} download>
          <QuestionMarkWrapper>?</QuestionMarkWrapper>
        </StyledHeaderHelp>
        <Link to={'/notifications'}>
          <Badge badgeContent={unreadCount} color={'info'}>
            <Notifications sx={{ color: 'white' }} />
          </Badge>
        </Link>
      </Box>
      <UserMenu />
    </StyledHeader>
  );
};

export { Header };
