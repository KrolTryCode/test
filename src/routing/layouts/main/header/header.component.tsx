import { Box, useMediaQuery, useTheme } from '@mui/material';
import { FC } from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';

import logo from '~/assets/logo/logo-small.png';
import { planPath } from '~/utils/configuration/routes-paths';

import { QuestionMarkWrapper, StyledHeader, StyledHeaderHelp } from './header.style';
import { NavMenuDesktop } from './nav-menu-desktop/nav-menu-desktop.components';
import { NavMenuMobile } from './nav-menu-mobile/nav-menu-mobile.component';
import { UserMenu } from './user-menu/user-menu.component';

interface HeaderProps {
  className?: string;
}

const Header: FC<HeaderProps> = ({ className }) => {
  const { t } = useTranslation();

  const theme = useTheme();

  const matches = useMediaQuery(theme.breakpoints.up('sm'));

  return (
    <StyledHeader className={className}>
      {matches ? (
        <>
          <Box margin={'auto 0'}>
            <Link to={planPath}>
              <img src={logo} alt={t('PROJECT_NAME')} />
            </Link>
          </Box>

          <NavMenuDesktop />
        </>
      ) : (
        <NavMenuMobile />
      )}
      <Box display={'flex'} flexDirection={'row'} marginLeft={'auto'}>
        <StyledHeaderHelp to={'/Help.pdf'} target={'_blank'} download>
          <QuestionMarkWrapper>?</QuestionMarkWrapper>
        </StyledHeaderHelp>
      </Box>
      <UserMenu />
    </StyledHeader>
  );
};

export { Header };
