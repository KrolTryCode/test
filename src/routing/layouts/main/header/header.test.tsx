import { screen } from '@testing-library/react';
import { t } from 'i18next';
import { it, describe, expect } from 'vitest';

import { renderComponent } from 'tests/utils/render-component';
import { useUserStore } from '~/app/user/user.store';

import { Header } from './header.component';

describe('Header', () => {
  beforeEach(() => {
    useUserStore.setState({
      data: {
        permissions: ['ADMIN'],
      },
    });

    renderComponent(<Header />, { withRouter: true });
  });

  // TODO mock redux userAuthorities
  it.skip('renders the header correctly', async () => {
    expect(await screen.findByRole('link', { name: '?' })).toBeInTheDocument();

    const container = await screen.findByTestId('appContainer');
    expect(container).toMatchSnapshot();
  });

  it('renders the header with admin permissions correctly', async () => {
    expect(await screen.findByRole('link', { name: '?' })).toBeInTheDocument();
    expect(await screen.findByRole('link', { name: t('NAVIGATION.ADMIN') })).toBeInTheDocument();
  });

  it('renders the header with user permissions correctly', async () => {
    useUserStore.setState({
      data: {
        permissions: ['USER'],
      },
    });

    expect(await screen.findByRole('link', { name: '?' })).toBeInTheDocument();
    expect(screen.queryByRole('link', { name: t('NAVIGATION.ADMIN') })).not.toBeInTheDocument();
  });
});
