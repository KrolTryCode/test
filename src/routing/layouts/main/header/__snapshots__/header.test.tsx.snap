// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders the header correctly 1`] = `
<div
  data-testid="appContainer"
>
  <header
    class="css-rdayrw"
  >
    <div
      class="MuiBox-root css-132diu7"
    >
      <a
        href="/plan"
      >
        <img
          alt="КГНЦ"
          src="/src/assets/logo/logo-small.svg"
        />
      </a>
    </div>
    <nav
      class="css-1wlw5c9"
    >
      <a
        class="css-1o7hvjg"
        href="/plan"
      >
        Текущая программа строительства
      </a>
      <a
        class="css-1o7hvjg"
        href="/future"
      >
        Перспективные потребности
      </a>
      <a
        class="css-1o7hvjg"
        href="/tasks-tree"
      >
        Дерево решений
      </a>
      <a
        class="css-1o7hvjg"
        href="/factories"
      >
        Карточки заводов
      </a>
      <a
        class="css-1o7hvjg"
        href="/classifier"
      >
        Классификаторы
      </a>
      <a
        class="css-1o7hvjg"
        href="/admin"
      >
        Администрирование
      </a>
      <a
        class="css-1o7hvjg"
        href="/reports"
      >
        Отчеты
      </a>
    </nav>
    <div
      class="MuiBox-root css-18j51vt"
    >
      <a
        class="css-1n1ivqf"
        download=""
        href="/Help.pdf"
        target="_blank"
      >
        <b
          class="css-18568zw"
        >
          ?
        </b>
      </a>
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-disableElevation  css-1qm1klz-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      id=":r0:"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiStack-root css-1rlitzi-MuiStack-root"
      >
        <div
          class="MuiStack-root css-163ljxj-MuiStack-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-jy59ac-MuiAvatar-root"
            color="secondary"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PersonOutlineOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5.9c1.16 0 2.1.94 2.1 2.1s-.94 2.1-2.1 2.1S9.9 9.16 9.9 8s.94-2.1 2.1-2.1m0 9c2.97 0 6.1 1.46 6.1 2.1v1.1H5.9V17c0-.64 3.13-2.1 6.1-2.1M12 4C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4m0 9c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4"
              />
            </svg>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1e3x312-MuiSvgIcon-root"
          data-testid="ExpandMoreRoundedIcon"
          focusable="false"
          stroke="currentColor"
          stroke-width="0.3"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.88 9.29 12 13.17 8.12 9.29a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59c.39-.39.39-1.02 0-1.41-.39-.38-1.03-.39-1.42 0"
          />
        </svg>
      </div>
    </button>
  </header>
</div>
`;
